modules/gapi/src/backends/oak/goakbackend.cpp

```diff
@@ -0,0 +1,616 @@
+                            cv::Size(static_cast<int>(oak_frame->getWidth()),
+                                     static_cast<int>(oak_frame->getHeight())),
+                            cv::gapi::oak::OAKFrameFormat::NV12,
+                            oak_frame->getData().data(),
+                            oak_frame->getData().data() + static_cast<long>(oak_frame->getData().size() / 3 * 2));
+            // FIXME: should we copy data instead?
+        } else {
+            GAPI_Assert(false && "Unsupported output type");
+        }
+
+        out.meta(out_arg, cv::util::get<cv::GRunArgs>(in_msg)[i].meta);
+        out.post(std::move(out_arg));
+    }
+}
+
+// Built-in kernels for OAK /////////////////////////////////////////////////////
+
+class GOAKBackendImpl final : public cv::gapi::GBackend::Priv {
+    virtual void unpackKernel(ade::Graph            &graph,
+                              const ade::NodeHandle &op_node,
+                              const cv::GKernelImpl &impl) override {
+        OAKGraph gm(graph);
+
+        const auto &kimpl  = cv::util::any_cast<cv::gimpl::GOAKKernel>(impl.opaque);
+        gm.metadata(op_node).set(cv::gimpl::OAKComponent{kimpl});
+    }
+
+    virtual EPtr compile(const ade::Graph &graph,
+                         const cv::GCompileArgs &args,
+                         const std::vector<ade::NodeHandle> &nodes,
+                         const std::vector<cv::gimpl::Data>& ins_data,
+                         const std::vector<cv::gimpl::Data>& outs_data) const override {
+        return EPtr{new cv::gimpl::GOAKExecutable(graph, args, nodes, ins_data, outs_data)};
+    }
+};
+
+cv::gapi::GBackend cv::gapi::oak::backend() {
+    static cv::gapi::GBackend this_backend(std::make_shared<GOAKBackendImpl>());
+    return this_backend;
+}
+
+namespace cv {
+namespace gimpl {
+namespace oak {
+
+// Kernels ///////////////////////////////////////////////////////////////
+
+template<class Impl, class K>
+class GOAKKernelImpl: public detail::OAKCallHelper<Impl, typename K::InArgs, typename K::OutArgs>
+                    , public cv::detail::KernelTag {
+    using P = detail::OAKCallHelper<Impl, typename K::InArgs, typename K::OutArgs>;
+public:
+    using API = K;
+    static cv::gapi::GBackend   backend() { return cv::gapi::oak::backend();  }
+    static GOAKKernel kernel()  { return GOAKKernel(&P::construct); }
+};
+
+#define GAPI_OAK_KERNEL(Name, API) \
+    struct Name: public cv::gimpl::oak::GOAKKernelImpl<Name, API>
+
+namespace {
+GAPI_OAK_KERNEL(GOAKEncFrame, cv::gapi::oak::GEncFrame) {
+    static void put(const std::unique_ptr<dai::Pipeline>& pipeline,
+                    const std::tuple<int, int>&,
+                    std::shared_ptr<dai::Node>& node,
+                    std::vector<std::pair<std::string, dai::Buffer>>&,
+                    GOAKContext::InputPtr* in,
+                    const cv::gapi::oak::EncoderConfig& cfg,
+                    GOAKContext::OutputPtr* out) {
```

Comment:
may be too much arguments here, I'd put everything which doesn't depend on `G_API_OP` definition to a separate structure and pass it along.

Reply:
Done. There is an idea to get rid of those at all (like hide pipeline/node type inside the macro) but it's hard to implement due to differences in DAI operations